[EC-654] Create commands for Group Create and Group Update

r-tome opened the pull request:
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
As part of work for [EC-449: Event Log - User for SCIM eventsDONE](https://bitwarden.atlassian.net/browse/EC-449) it was identified the necessity for breaking up the GroupService.SaveAsync method logic into two commands - one for creating a group and another for updating.


## Code changes

* **bitwarden_license/src/Scim/Utilities/ScimServiceCollectionExtensions.cs:** Removed injection of `IDeleteGroupCommand` as it is already being added on `OrganizationServiceCollectionExtensions`
* **src/Core/OrganizationFeatures/Groups/CreateGroupCommand.cs:** Implementation of the command responsible of creating a group in the repository
* **src/Core/OrganizationFeatures/Groups/Interfaces/ICreateGroupCommand.cs:** Interface of the command responsible of creating a group in the repository
* **src/Core/OrganizationFeatures/Groups/Interfaces/IUpdateGroupCommand.cs:** Interface of command responsible of updating a group in the repository
* **src/Core/OrganizationFeatures/Groups/UpdateGroupCommand.cs:** Implementation of command responsible of updating a group in the repository
* **src/Core/OrganizationFeatures/OrganizationServiceCollectionExtensions.cs:** Added injection of group commands
* **src/Core/Services/Implementations/GroupService.cs:** Removed logic and replaced it with calling each respective command
* **test/Core.Test/OrganizationFeatures/Groups/CreateGroupCommandTests.cs:** Testing CreateGroupCommand
* **test/Core.Test/OrganizationFeatures/Groups/UpdateGroupCommandTests.cs:** Testing UpdateGroupCommand
* **test/Core.Test/Services/GroupServiceTests.cs:** Updated tests to check that the commands are being called

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

eliykat commented on test/Core.Test/OrganizationFeatures/Groups/UpdateGroupCommandTests.cs:
```diff
@@ -0,0 +1,61 @@
+    [BitAutoData]
+    public async Task UpdateGroup_Success(SutProvider<UpdateGroupCommand> sutProvider, Group group)
+    {
+        sutProvider.GetDependency<IGroupRepository>()
+            .GetByIdAsync(group.Id)
+            .Returns(group);
```
Also seems unnecessary (and in all other test methods in this file)

eliykat reviewed (changes requested):
I love how simple these tests are now

r-tome commented on test/Core.Test/OrganizationFeatures/Groups/CreateGroupCommandTests.cs:
```diff
@@ -0,0 +1,67 @@
+
+        await sutProvider.GetDependency<IGroupRepository>().Received(1).CreateAsync(group);
+        await sutProvider.GetDependency<IEventService>().Received(1).LogGroupEventAsync(group, Enums.EventType.Group_Created);
+        Assert.True(group.CreationDate - utcNow < TimeSpan.FromSeconds(1));
```
> This looks flaky! It makes me think we should wrap `DateTime` in a service so that we can mock it for testing. Out of scope here though, just thought I'd flag it.

I didn't find a simple way of doing this without modifying every single usage of DateTime to use a new custom interface.

eliykat replied:
Yes, I think that's what you'd have to do, which is why I think it's out of scope here.

r-tome commented:
@eliykat, thanks for the suggestions. I've replaced `GroupService.SaveAsync` with the commands `CreateGroupCommand` and `UpdateGroupCommand` and cleaned up the unit tests.

eliykat commented on bitwarden_license/src/Scim/Groups/PostGroupCommand.cs:
```diff
@@ -39,7 +39,7 @@ public async Task<Group> PostGroupAsync(Guid organizationId, ScimGroupRequestMod
 
         var group = model.ToGroup(organizationId);
-        await _groupService.SaveAsync(group, EventSystemUser.SCIM, null);
+        await _createGroupCommand.CreateGroupAsync(group, EventSystemUser.SCIM, null);
```
Are we sure that SCIM only uses POST for creating and not updating? Can you please check the spec? (and/or the spec for the specific IdPs we support)

r-tome replied:
Apparently it can be used for updating on some endpoints (https://www.rfc-editor.org/rfc/rfc7644.html#section-3.2):
> POST   Depending on the endpoint, creates new resources, creates a
>           search request, or MAY be used to bulk-modify resources.

The only endpoints I see that use that feature are [Bulk](https://www.rfc-editor.org/rfc/rfc7644.html#section-3.7) and [Search](https://www.rfc-editor.org/rfc/rfc7644.html#section-3.4.3).

In the case of the Users and Groups endpoints they are only used for creating https://www.rfc-editor.org/rfc/rfc7644.html#section-3.3

eliykat commented on bitwarden_license/src/Scim/Groups/PatchGroupCommand.cs:
```diff
@@ -49,15 +53,15 @@ public async Task PatchGroupAsync(Guid organizationId, Guid id, ScimPatchModel m
                 else if (operation.Path?.ToLowerInvariant() == "displayname")
                 {
                     group.Name = operation.Value.GetString();
-                    await _groupService.SaveAsync(group, EventSystemUser.SCIM);
+                    await _updateGroupCommand.UpdateGroupAsync(group, EventSystemUser.SCIM);
```
I don't think we're supposed to call commands from other commands. But I don't know a better solution for now without revisiting this command entirely, so I think it's fine for now.

eliykat commented on src/Core/Services/Implementations/GroupService.cs:
```diff
@@ -13,93 +12,23 @@ public class GroupService : IGroupService
     private readonly IOrganizationUserRepository _organizationUserRepository;
     private readonly IGroupRepository _groupRepository;
-    private readonly IReferenceEventService _referenceEventService;
+    private readonly ICreateGroupCommand _createGroupCommand;
+    private readonly IUpdateGroupCommand _updateGroupCommand;
```
These are now unused in this class.

eliykat commented on src/Core/OrganizationFeatures/Groups/CreateGroupCommand.cs:
```diff
@@ -0,0 +1,71 @@
+
+    private async Task GroupRepositoryCreateGroupAsync(Group group, IEnumerable<SelectionReadOnly> collections = null)
+    {
+        var organization = await _organizationRepository.GetByIdAsync(group.OrganizationId);
+        if (organization == null)
+        {
+            throw new BadRequestException("Organization not found");
+        }
```
We can't have database reads in a command. I can see why you changed this (to remove the extraneous parameter), but it's out of place here.

On the other hand, requiring the `Organization` object to be passed in just to check its `Id` (when we're not otherwise doing anything with it) seems a bit pointless. I'm inclined to remove it altogether, or move it up to the controller level.

Could you please check with @MGibson1?

MGibson1 replied:
yup, I’d move that up to the controller to grab the org, create another method on the command class (`Validate` or some such) and use that method to check the org exists and can use groups

eliykat replied:
Right, I missed the `CanUseGroups` check. Makes sense, ty

r-tome replied:
Thank you both for your suggestions. I've extracted the database read from the command to the respective controllers and added unit tests around them.

eliykat commented on src/Core/OrganizationFeatures/Groups/UpdateGroupCommand.cs:
```diff
@@ -0,0 +1,65 @@
+
+    private async Task GroupRepositoryUpdateGroupAsync(Group group, IEnumerable<SelectionReadOnly> collections = null)
+    {
+        var organization = await _organizationRepository.GetByIdAsync(group.OrganizationId);
+        if (organization == null)
+        {
+            throw new BadRequestException("Organization not found");
+        }
```
Same here re: not mixing read & write operations.

eliykat commented on test/Core.Test/OrganizationFeatures/Groups/CreateGroupCommandTests.cs:
```diff
@@ -0,0 +1,83 @@
+    [Theory, OrganizationCustomize(UseGroups = true), BitAutoData]
+    public async Task CreateGroup_Success(SutProvider<CreateGroupCommand> sutProvider, Organization organization, Group group)
+    {
+        sutProvider.GetDependency<IOrganizationRepository>().GetByIdAsync(group.OrganizationId).Returns(organization);
```
This might be removed depending on what we do with my feedback above. Just flagging it.

eliykat commented on test/Core.Test/OrganizationFeatures/Groups/CreateGroupCommandTests.cs:
```diff
@@ -0,0 +1,83 @@
+
+        await sutProvider.Sut.CreateGroupAsync(group);
+
+        await sutProvider.GetDependency<IGroupRepository>().Received(1).CreateAsync(group);
+        await sutProvider.GetDependency<IEventService>().Received(1).LogGroupEventAsync(group, Enums.EventType.Group_Created);
```
I think you should also assert the call to `ReferenceEventService` in each happy path test.

eliykat commented on test/Api.Test/Controllers/GroupsControllerTests.cs:
```diff
@@ -0,0 +1,143 @@
+
+[ControllerCustomize(typeof(GroupsController))]
+[SutProviderCustomize]
+public class GroupsControllerTests
```
FYI - I don't think we often unit test controllers, because they should just call other classes which have the important business logic in them (services/commands/queries). So we test those other classes instead. The controllers still have some logic in them, but they should be very simple, "flat" (non-nested) logic which can live without a unit test.

This is partly what I'm getting at in my other comments on this class - you end up doing quasi-integration tests, or indirectly testing dependencies, or testing not much at all. The happy path tests are fine, although I wouldn't require them, and I think we can remove the others. Just my 2 cents, tell me if you disagree.

eliykat commented on test/Api.Test/Public/Controllers/GroupsControllerTests.cs:
```diff
@@ -0,0 +1,132 @@
+
+[ControllerCustomize(typeof(GroupsController))]
+[SutProviderCustomize]
+public class GroupsControllerTests
```
My comments above for the other `GroupsControllerTests` also apply here.

r-tome commented:
Thanks for the review, @eliykat. I've changed the `Validate` methods to have a private accessor and removed the unnecessary Controller unit tests that you pointed out. Your explanation on what we should test and where is really helpful as I have a tendency to test everything everywhere 😅
